### PR TITLE
[SYCL] Update tests with new test cases for FPGA function attributes 

### PR DESCRIPTION
**Pull request by smanna12**
Template parameter support was added for

 1. [[intel::max_global_work_dim)]] attribute on https://github.com/intel/llvm/pull/2816
 2. [[intel::num_simd_work_items()]] attribute on https://github.com/intel/llvm/pull/2510
 3. [[intel::reqd_sub_group_size()]] attribute on https://github.com/intel/llvm/pull/1807

This patch adds the following new test cases that were not there before to verify/improve the template support:

 1. Test that checks wrong function template instantiation and ensures that the type
    is checked properly when instantiating from the template definition.
 2. Test that checks expression is not a constant expression.
 3. Test that checks expression is a constant expression.
 4. Test that checks template parameter support on function.

NOTE: No change in compiler. All new test cases have already been supported.

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

File: clang/test/SemaSYCL/sycl-device-num_simd_work_items-template.cpp [hunk 1]
```diff
@@ -2,6 +2,28 @@
...
+// Test that checks expression is not a constant expression.
+int foo();
+// expected-error@+1{{'num_simd_work_items' attribute requires an integer constant}}
+[[intel::num_simd_work_items(foo() + 12)]] void func1();
+
+// Test that checks expression is a constant expression.
+constexpr int bar() { return 0; }
+[[intel::num_simd_work_items(bar() + 12)]] void func2(); // OK
+
+// Test that checks template parameter suppport on member function of class template.
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/sycl-device-num_simd_work_items-template.cpp:26]: support*

File: clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp [hunk 2]
```diff
@@ -2,6 +2,28 @@
...
+// Test that checks expression is not a constant expression.
+int foo();
+// expected-error@+1{{'max_global_work_dim' attribute requires an integer constant}}
+[[intel::max_global_work_dim(foo() + 1)]] void func1();
+
+// Test that checks expression is a constant expression.
+constexpr int bar() { return 0; }
+[[intel::max_global_work_dim(bar() + 2)]] void func2(); // OK
+
+// Test that checks template parameter suppport on member function of class template.
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp:26]: support*

File: clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp [hunk 3]
```diff
@@ -2,6 +2,28 @@
...
+// Test that checks expression is not a constant expression.
+int foo();
+// expected-error@+1{{'reqd_sub_group_size' attribute requires an integer constant}}
+[[intel::reqd_sub_group_size(foo() + 12)]] void func1();
+
+// Test that checks expression is a constant expression.
+constexpr int bar() { return 0; }
+[[intel::reqd_sub_group_size(bar() + 12)]] void func2(); // OK
+
+// Test that checks template parameter suppport on member function of class template.
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp:26]: support*

File: clang/test/SemaSYCL/sycl-device-num_simd_work_items-template.cpp (hunk 1 above)
Comment (smanna12) [reply clang/test/SemaSYCL/sycl-device-num_simd_work_items-template.cpp]: Sorry for the typo. Fixed now

File: clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp (hunk 2 above)
Comment (smanna12) [reply clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp]: Done

File: clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp (hunk 3 above)
Comment (smanna12) [reply clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp]: Done

**Review by elizabethandrews — APPROVED** [review]
LGTM. Thanks!

**Comment by smanna12** [comment]
> LGTM. Thanks!

Thanks @elizabethandrews for the reviews.

File: clang/test/CodeGenSYCL/intel-max-global-work-dim.cpp [hunk 4]
```diff
@@ -26,12 +29,17 @@ int main() {
 
     Functor<2> f;
     h.single_task<class kernel_name3>(f);
+
+    h.single_task<class kernel_name4>([]() {
+      func<2>();
```
Comment (AaronBallman) [inline clang/test/CodeGenSYCL/intel-max-global-work-dim.cpp:34]: Can you also ensure there's a test for an instantiation like `func<-1>();`? If the test is expected to cause errors, it should be a SemaSYCL test, but if the code is accepted, there should be both a SemaSYCL test (for the warning) and a CodeGenSYCL test (for the codegen).

File: clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp [hunk 5]
```diff
@@ -2,6 +2,28 @@
...
+// Test that checks wrong function template instantiation and ensures that the type
+// is checked properly when instantiating from the template definition.
+template <typename Ty>
+// expected-error@+1{{'max_global_work_dim' attribute requires an integer constant}}
+[[intel::max_global_work_dim(Ty{})]] void func() {}
+
+struct S {};
+void var() {
+  //expected-note@+1{{in instantiation of function template specialization 'func<S>' requested here}}
+  func<S>();
```
Comment (AaronBallman) [inline clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp:14]: Another interesting test would be: `func<float>();` to see if we get an exciting implicit conversion to int there or not. Also, a test like `func<int>();`, which should be accepted.

File: clang/test/CodeGenSYCL/intel-max-global-work-dim.cpp (hunk 4 above)
Comment (smanna12) [reply clang/test/CodeGenSYCL/intel-max-global-work-dim.cpp:34]: Thanks @AaronBallman for looking into the patch.

>>Can you also ensure there's a test for an instantiation like func<-1>();
Test is added

All attributes are expected to cause errors. So SemaSYCL tests are added here.
[intel::max_global_work_dim)]] 
[[intel::num_simd_work_items()]] 
[[intel::reqd_sub_group_size()]]

File: clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp (hunk 5 above)
Comment (smanna12) [reply clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp:14]: >>Another interesting test would be: func<float>(); to see if we get an exciting implicit conversion to int there or not.

Test is added. We get an implicit conversion to int for all attributes here.

>>Also, a test like func<int>();, which should be accepted.

Test is added.
[intel::max_global_work_dim)]] -- only accept this test
[[intel::num_simd_work_items()]] -- does not accept this test 
[[intel::reqd_sub_group_size()]] -- does not accept this test

Comment (AaronBallman) [reply clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp:14]: Is it expected that we get an implicit conversion for `func<float>()` or was that a surprise (if a surprise, maybe open an issue for it)?

Also, is the inconsistency between the other attributes with `func<int>()` expected? I would assume we'd want all of these or none of these to behave the same way. If it's unexpected, maybe open another issue?

**Review by AaronBallman — APPROVED** [review]
Changes LGTM!

File: clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp (hunk 5 above)
Comment (smanna12) [reply clang/test/SemaSYCL/sycl-device-intel-max-global-work-dim-template.cpp:14]: I have created https://github.com/intel/llvm/issues/3100 for this issue and assigned to me. I will investigate this behavior for all of these in a separate PR. Thanks @AaronBallman for suggesting about this tests.

**Review by elizabethandrews — APPROVED** [review]

**Comment by smanna12** [comment]
Thank you @elizabethandrews and @AaronBallman for the discussion and reviews.

**Review by premanandrao — APPROVED** [review]
Sorry, I missed the earlier notifications.
Looks good to me.

**Comment by smanna12** [comment]
> Sorry, I missed the earlier notifications.
> Looks good to me.

Thanks @premanandrao.

**Pull request merged**